File: parameter.jl
mutable struct Parameter <: Comparable
    ref::String
    name::String
    in::String
    description::String
    required::Bool
    deprecated::Bool
    allowEmptyValue::Bool
    style::String
    explode::Bool
    allowReserved::Bool
    schema::Schema
    example::Any
    examples::Dict{String,Example}
    content::Dict{String,MediaType}
end

Parameter() = Parameter("", "", "", "", false, false, false, "", false, false, Schema(), Nothing,
    Dict{String,Example}(), Dict{String,MediaType}())

function parse!(p::Parameter, data::Dict{Any,Any})
    for (key, value) in data
        if key == "\$ref" p.ref = value end
        if key == "name" p.name = value end
        if key == "in" p.in = value end
        if key == "description" p.description = value end
        if key == "required" p.required = value end
        if key == "deprecated" p.deprecated = value end
        if key == "allowEmptyValue" p.allowEmptyValue = value end
        if key == "style" p.style = value end
        if key == "explode" p.explode = value end
        if key == "allowReserved" p.allowReserved = value end
        if key == "schema" parse!(p.schema, value) end
        if key == "example" p.example = value end
        if key == "examples" parse!(p.examples, value) end
        if key == "content" parse!(p.content, value) end
    end
end

function parse!(v::Vector{Parameter}, data::Vector{Dict{Any,Any}})
    for d in data
        p = Parameter()
        parse!(p, d)
        push!(v, p)
    end
end

File: src/parser/parameter.cpp
//
// Created by Rakesh on 19/09/2025.
//

#include "parser.hpp"
#include "model/parameter.hpp"

template <>
void spt::parser::parse( model::Parameter& m, c4::yml::ConstNodeRef node )
{
  for ( const auto& child : node.children() )
  {
    if ( child.key() == "$ref" ) child >> m.ref;
    if ( child.key() == "name" ) child >> m.name;
    if ( child.key() == "in" ) child >> m.in;
    if ( child.key() == "description" ) child >> m.description;
    if ( child.key() == "style" ) child >> m.style;
    if ( child.key() == "required" ) child >> m.required;
    if ( child.key() == "deprecated" ) child >> m.deprecated;
    if ( child.key() == "allowEmptyValue" ) child >> m.allowEmptyValue;
    if ( child.key() == "explode" ) child >> m.explode;
    if ( child.key() == "allowReserved" ) child >> m.allowReserved;
    if ( child.key() == "example" && child.has_val() ) m.example = std::string{ child.val().begin(), child.val().end() };
    if ( child.key() == "examples" )
    {
      for ( const auto& ex : child.children() )
      {
        auto key = std::string_view{ ex.key() };
        m.examples.try_emplace( std::string{ key }, parse<model::Example>( ex ) );
      }
    }
    if ( child.key() == "schema" ) m.schema = parse<model::Schema>( child );
  }
}
